htlcswitch/link_test.go

```diff
@@ -0,0 +1,626 @@
+
+		// Log messages that bob receives from carol.
+		n.bobServer.record(createLogFunc("alice<-[bob]<-carol: ",
+			n.secondBobChannelLink.ChanID()))
+
+		// Log messages that carol receives from bob.
+		n.carolServer.record(createLogFunc("alice->bob->[carol]",
+			n.carolChannelLink.ChanID()))
+	}
+
+	var amount btcutil.Amount = btcutil.SatoshiPerBitcoin
+	errChan, invoice, err := n.makePayment(n.carolServer, amount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for:
+	// * HTLC add request to be sent from Alice to Bob.
+	// * Alice<->Bob commitment states to be updated.
+	// * HTLC add request to be propagated to Carol.
+	// * Bob<->Carol commitment state to be updated.
+	// * settle request to be sent back from Carol to Bob.
+	// * Alice<->Bob commitment state to be updated.
+	// * settle request to be sent back from Bob to Alice.
+	// * Alice<->Bob commitment states to be updated.
+	// * user notification to be sent.
+	select {
+	case err := <-errChan:
+		if err != nil {
+			t.Fatalf("something wrong went when sending request: "+
+				"%v", err)
+		}
+		break
+	case <-time.After(time.Second):
+		t.Fatal("htlc was no settled in time")
+	}
+
+	// Wait for Bob to receive the revocation.
+	time.Sleep(100 * time.Millisecond)
+
+	// Check that Carol invoice was settled and bandwidth of HTLC
+	// links were changed.
+	if !invoice.Terms.Settled {
+		t.Fatal("alice invoice wasn't settled")
+	}
+
+	if aliceBandwidthBefore-amount != n.aliceChannelLink.Bandwidth() {
+		t.Fatal("the bandwidth of alice channel link which handles " +
+			"alice->bob channel wasn't decreased on htlc amount")
+	}
+
+	if firstBobBandwidthBefore+amount != n.firstBobChannelLink.Bandwidth() {
+		t.Fatal("the bandwidth of bob channel link which handles " +
+			"alice->bob channel wasn't increased on htlc amount")
+	}
+
+	if secondBobBandwidthBefore-amount != n.secondBobChannelLink.Bandwidth() {
+		t.Fatal("the bandwidth of bob channel link which handles " +
+			"bob->carol channel wasn't decreased on htlc amount")
+	}
+
+	if carolBandwidthBefore+amount != n.carolChannelLink.Bandwidth() {
+		t.Fatal("the bandwidth of carol channel link which handles " +
+			"carol->bob channel wasn't decreased on htlc amount")
+	}
+}
+
+// TestChannelLinkMultiHopInsufficientPayment checks that we receive error if
+// bob<->alice channel has insufficient BTC capacity/bandwidth. In this test we
+// send the payment from Carol to Alice over Bob peer. (Carol -> Bob -> Alice)
+func TestChannelLinkMultiHopInsufficientPayment(t *testing.T) {
+	n := newThreeHopNetwork(t)
+	if err := n.start(); err != nil {
+		t.Fatalf("can't start three hop network: %v", err)
+	}
+	defer n.stop()
+
+	carolBandwidthBefore := n.carolChannelLink.Bandwidth()
+	firstBobBandwidthBefore := n.firstBobChannelLink.Bandwidth()
+	secondBobBandwidthBefore := n.secondBobChannelLink.Bandwidth()
+	aliceBandwidthBefore := n.aliceChannelLink.Bandwidth()
+
+	var amount btcutil.Amount = 4 * btcutil.SatoshiPerBitcoin
```

Comment:
Perhaps either this capacity of all the channels in the 3 hop network should be exposed, or be taken in as a parameter to the constructor `newThreeHopNetwork`?